### DISCUSSION TITLE
Inherit Constraints with Classes

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#3810] [Language Ideas] [answered]
Today, you do not have to explicitly specify constraints on overridden methods.  The following is valid:
````
    public class Foo1 {
        public virtual void DoSomething<T>() where T:Extension { }
    }

    public class Bar1 : Foo1 {
        public override void DoSomething<T>() { //No respecification necessary
            base.DoSomething<T>();
        }
    }
````

However, when working with classes, you have to:
````
    public class Foo2<T> where T : Extension { }

    public class Bar2<T> : Foo2<T> { } //This causes an error.  I have to specify where T:Extension
````

Please allow derived types to automatically inherit type parameter constraints in the same way that methods do.

**Comment by CyrusNajmabadi** [comment]
With overrides you are restating the exact same signature.  It is the same type parameter.  It's not that you're inheriting by default.  It's that they literally must be the same, so forcing you to have to write this would be onerous.  You also end up in scenarios that are literally not legal to write in terms of the grammar.  for example:

```c#
class C<X>
{
    public virtual void Foo<T>() where T : X { }
}

class D : C<int>
{
    public override void Foo<T>() // where T : int ??
}
```

this cannot be written/expressed in the language, so it wouldn't be possible to even restate the constraints even if we required that.

With classes this is not the same.  With classes they are *not* the same type parameters.  The type parameters of the derived type are distinct and unrelated to those of the base type.

**Comment by TonyValenti** [comment]
@CyrusNajmabadi -
Yes, but while they're different, they should be able to automatically inherit the constraints based on how they are used.

For example:
````

public class B<T> where T  : Extension {}

public public class C<TKey, TValue> : B<TKey> {

}
````

That should cause ````TKey```` to inherit all of the ````T```` constraints because it is used as a value for T.

**Comment by CyrusNajmabadi** [comment]
i very much do not like that because this is a declaration position and C appears to not have any constraints here. Furthermore, i don't see why this class/base relationship would be special.  Say you had:

```c#
public public class C<TKey, TValue>  {
    public B<TKey> B = new B<TKey>(); // should TKey now implicitly gain all the constraints of B's type parameter?
}
```

The override case is different in that, as i specified above:

1. it literally must be the same.
2. there is no syntax to specify constraints legally for many things that would occur.
3. you cannot actually call overrides directly. you call the virtual.  

in other words, there is a strict and exact overlap with virtual/override. no such relation exists with classes.

**Comment by ZacharyPatten** [comment]
You don't always want the same constraints on derived class' generics. For example you may want a tighter constraint than the base:
```cs
class A<T1> where T1 : IList { }
class B<T1> : A<T1> where T1 : List<string> { }
```

Or you may need one generic parameter to be dependent on another generic parameter:
```cs
public class A<T1> where T1 : IList { }
public class B<T1, T2> : A<T2> where T2 : List<T1> { }
```

**Comment by mburbea** [comment]
I personally wish that you could redeclare the method constraints again when it is valid and legal to do so for readability purposes.
https://github.com/dotnet/csharplang/issues/353

It's gotten me confused more than once when implementing an override that I might forget that I have `T:new()` or that `T` extends some interface.

**Comment by IS4Code** [comment]
This seems like something a source generator could also do.

**Comment by Joe4evr** [comment, marked as answer]
See also: https://ericlippert.com/2012/03/09/why-not-automatically-infer-constraints/

**Comment by iskiselev** [comment]
There are few more same issue in old roslyn repo:
dotnet/roslyn#8041, dotnet/roslyn#8304

**Comment by tuscen** [comment]
It would still be a very handy feature. It could be constrained in scope here and there to make it easier to implement and reason about, though I don't know where and how exactly.

**Comment by Xyncgas** [comment]
Maybe we should try it out, see if we like it, then say it's a bad thing, you can't talk your way out of eating vegetables, if you had never tried it
To quote Lippert : " C# is not a “figure out what the developer meant to say and say it for them” language; it’s a “tell the developer when they have supplied too little information” language"
